specs/PushNotifications/ToastNotifications-spec.md

```diff
@@ -0,0 +1,551 @@
+stack.
+
+**Registration**: The Windows SDK today provides AppId overloads to developers who can specify their
+own Identifier for Unpackaged apps. The platform is currently designed to create an explicit
+Registration for this overloaded AppId which causes all sorts of problems (string size limits being
+hit, Personally Identifiable Information stored as part of AppIds, Registrations tagged with
+incorrect App Type etc). The WinApp SDK gets rid of the AppId overloads in the Public APIs.
+
+**Activation**: Setting up activation handlers using Background Triggers is challenging for
+developers because we don't have downlevel support for all OS SKUs. Moreover, the steps for setting
+up activation handlers vary greatly for different app types MSIX Vs Unpackaged Vs UWP. We will need
+to abstract away the activation technology and simplify this process for developers for all
+supported downlevel OS SKUs.
+
+# Description
+
+This API enables all Win32 applications to display app notifications irrespective
+of their app type. This includes unpackaged apps and packaged win32 (MSIX Desktop Bridge, MSIX
+Win32App, Sparse Signed Packages). Moreover, all scenarios should adhere to OS resource
+management policies like Power Saver, Network Attribution (amount of data an app uses), enterprise
+group policies, etc. The WinAppSDK will abstract away the complexities of
+dealing with Notification delivery and related activations as much as possible freeing the developer
+to focus on other app related challenges.
+
+The App Notification API does the following:
+
+-   Allow Unpackaged apps to register for notifications.
+-   Allow Packaged apps to register for notifications.
+-   Allow Unpackaged apps to display cloud sourced notifications.
+-   Allow packaged apps to display cloud sourced notifications.
+-   Allow packaged apps to display local notifications
+-   Allow unpackaged apps to display local notifications
+-   Allow packaged apps to Update/Delete their local notifications.
+-   Allow unpackaged apps to Update/Delete their local notifications.
+-   Allow packaged applications to process notification activators.
+-   Allow unpackaged apps to process notification activators.
+
+# Examples
+
+## Packaged App Registering and handling a notification
+
+This scenario is specifically geared towards Packaged apps that need to send an app notification. An app
+Developer Registers their app for notifications and sets it up for background\foreground
+activation. Adding a AppNotificationActivated event handler signals that the app is ready to accept and handle
+an action in the foreground where the user clicks on a Notification in the action centre or a UI element in
+the payload body (like a Button). The developer can inturn run some business logic in their app in
+response to the UI Action.
+
+```cpp
+
+CLSID const c_comServerId = { 0x43136EB5,0xD36C,0x11CF,{0xAD,0xBC,0x00,0xAA,0x00,0xA8,0x00,0x33} };
```

Comment:
*(And if it's too slow/intensive to parse their manifest, we should consider other ways to validate whether the value is right, Brendan suggested we could look up to see if a COM server is registered with the corresponding CLSID... the important thing to remember is to **provide instantaneous, meaningful feedback to the developer when they make a mistake**... otherwise if they make a typo (which they will), they'll be utterly perplexed why their notifications are silently failing... and developers have wasted hours debugging issues similar to this in our APIs since we don't provide meaningful, immediate errors)*

Reply:
Yes. This was our POR. We will figure out the COM CLSID and validate the launch process as part of the Registration flow.

Reply:
We figured out a way to infer to the Registry and avoid manifest overhead for now.